### ISSUE TITLE
exports lib packages in package.json

### ISSUE THREAD

**Issue by trim21** [axios/axios#5000]
TLDR：

> Oh, yeah I don't think we will ever export that.

https://github.com/axios/axios/issues/5000#issuecomment-1311350960

#### Is your feature request related to a problem? Please describe.

Yes. I'm writing a adaptor to be used in userscript like tampermonkey or Violentmonkey, 
it use a xhr provided be broser extension `GM.xmlHttpRequest` to make cors request.

And to make it behavior like axios' xhr adaptor, I'm using some axios' helper functions directly by 

```ts
import buildFullPath from "axios/lib/core/buildFullPath";
import settle from "axios/lib/core/settle";
import buildURL from "axios/lib/helpers/buildURL";
import parseHeaders from "axios/lib/helpers/parseHeaders";
import utils from "axios/lib/utils";
```

But after axios **1.0.0** is released, a `exports` field is added to package.json, and lib directory is no longer exported anymore, this make me unable to use helper function provided by axios anymore.

https://github.com/axios/axios/blob/484aa4fe6addccbd32206a31eb3d2d000843066e/package.json#L6-L17

modern js bundler will raise a error for this: `Package subpath './lib/core/buildFullPath' is not defined by "exports" in .../node_modules/axios/package.json` by bundler.

#### Describe the solution you'd like

add `lib` dir to `package.json#exports`

**Comment by arthurfiorette** [comment]
And, if I'm not mistaken, react native users need that the package.json field to be exported as well.

Some users of axios-cache-interceptors raised an issue about this same problem and after some reading, this may also be the best way to export an project that is coded by default with ESM:

https://github.com/arthurfiorette/axios-cache-interceptor/blob/4bf049b935d47f5fc46c587ff81c0e4d15b52e8b/package.json#L9-L20

**Comment by arthurfiorette** [comment]
Also, it would be great if a simple `export * from '../index.d.ts` declaration file was shipped with the same name for each file in the dist folder, this way direct imports would work flawlessly with typescript.

Like: `import ... from 'axios/dist/node/axios.cjs'`

**Comment by trim21** [comment]
No, this is not fixed by #5030, it's a complete different problem.

**Comment by trim21** [comment]
@jasonsaayman could you re-open this issue?

**Comment by ivanpepelko** [comment]
> No, this is not fixed by #5030, **it's a complete different problem**.
> 
> People who are affected by this issue want to import utils from `axios/lib/...`, which is curretly forbidden by `package.json#exports`.
> 
> Currently `package.json#exports` only allow us to `import 'axios'`, but we are not just importing `axios`, we are importing `axios/lib/...`, and change exports of `axios` **won't help**.
> 
> The other issues closed by this #5030 are indeed importing from `axios`, they are fixed by #5030, but not this one.

This is correct, I made a mistake in when describing #5030, as this issue was mentioned in https://github.com/axios/axios/issues/5008#issuecomment-1268735288.

**Comment by tusbar** [comment]
Related: https://github.com/jamesmbourne/aws4-axios/issues/701

**Comment by kranzhoff** [comment]
I think my issue is related to this:
Trying to set the axios.defaults.adapter in my TS project's test files like this:

`axios.defaults.adapter = require('axios/lib/adapters/http')`

fails with axios 1.x due to the httpAdapter not being exported.
This used to work just fine with axios 0.27.2.

**Comment by jasonsaayman** [comment]
Hi :wave:

Please try the latest pre-release by running the following:

```bash
npm i axios@1.2.0-alpha.1
```

Please provide feedback in either the pinned issue or the discussion thread 🧵

**Comment by trim21** [comment]
I'm 70% sure it's not what we want… utils are still not exported, makes no difference

**Comment by jasonsaayman** [comment]
Oh, yeah I don't think we will ever export that. Exporting the entire libs would cause a large potential issue and cause people to use Axios outside of the scope of its declared public API. If you could motivate what you would like and why maybe I would look into it but blanket exports of the directory would not be smart.

**Comment by trim21** [comment]
Ok

**Comment by jasonsaayman** [comment]
Yeah sorry, I get what you are trying to do but should we do this we would then open ourselves up to so much more issues if we declared this as part of the public API. That being said if you would like to roll what you are working on into a package we can maybe work on that and see how we can make it possible to provide this to a consumer.

**Comment by trim21** [comment]
I think it's very understandable that these API are not stable.You can document that they may change any time and export them.

**Comment by florianbepunkt** [comment]
@jasonsaayman One possible use case is aws4-axios, which uses

```
import buildUrl from "axios/lib/helpers/buildURL";
import combineURLs from "axios/lib/helpers/combineURLs";
import isAbsoluteURL from "axios/lib/helpers/isAbsoluteURL";
```

A possible solution for package maintainers would be to duplicate these axios methods in their own codebase.

On the other hand I believe the majority of signing-related middleware/interceptors would need these set of helpers.  Do you think increasing the surface of the public API will increase issues in general, or is there anything particular about these helpers (e. g. frequent changes).

**Comment by dhensby** [comment]
Still doesn't work for me. 

`Error [ERR_PACKAGE_PATH_NOT_EXPORTED]: Package subpath './lib/core/settle' is not defined by "exports" in .\node_modules\axios\package.json`

The [docs here](https://github.com/axios/axios/blob/v1.x/lib/adapters/README.md) are quire explicit about using the `settle.js` function both in testing and if creating bespoke adapters.

**Comment by spyndutz** [comment]
I'm not sure if this is the correct suggestion since it's basically changing how Axios package is being shipped (And the effort is definitely huge). What if we modularize it and make the repository a monorepo approach similar to what [parcel](https://github.com/parcel-bundler/parcel) did?  
So we have scoped package like `@axios/core`, `@axios/helpers`, `@axios/adapters` and let them import from those instead? While the main Axios package itself only export the public API it intended, so we have a different package versioning for those scpoed modules and the main Axios package.

**Comment by guoyunhe** [comment]
@jasonsaayman could you please provide a way to export built-in adaptors, which is a config option of axios? I want to use http adaptor in jsdom environment. jsdom's XHR implementation is a trouble...

**Comment by DigitalBrainJS** [comment]
@guoyunhe Manual adapter switching without the need for direct import will be provided by #5277 PR.

**Comment by bigmeow** [comment]
I have a solution and it wasted me a few days: 
make it not work  that  "exports"  field in the package.json file, see the [webpack doc](https://webpack.js.org/configuration/resolve/#resolveexportsfields).

### webpack.config.js
```js
module.exports = {
  //...
  resolve: {
    exportsFields: ['exports2'],
  },
};
```
### or webpackChain
```
chain.merge({
    resolve: {
      exportsFields: ["exports2"],
    }
});
```

It really works

@trim21  @kranzhoff  @guoyunhe

**Comment by arthurfiorette** [comment]
@DigitalBrainJS, Actually, this just extends the current problem into smaller ones. Everytime someone will be needing an internal function from axios that is not exported. As you can see in #5324, #5264 and so on...

Why can't all functions be exported as normal so anyone can use it while integrating with axios? The bytesize wouldn't be that big? or would it?

**Comment by jasonsaayman** [comment]
@arthurfiorette we don't want to export everything else people will rely on code that can change without being documented as we can change underlying implementations without changing the public API. The only way I can see this working is we either export it and then whoever uses it does so explicitly at their own risk or we export only what is needed. For now I would like to do it one by one but if there are many uses cases and people really want us to do so we can look into everything being exported.

**Comment by damirbogdanov** [comment]
the following fixed the issue for me

### webpack.config.js

```js
module.exports = {
  resolve: {
    alias: {
      "axios/lib": path.resolve(__dirname, "../../node_modules/axios/lib"),
    },
  },
};

```

**Comment by Liubasara** [comment]
any news? same problem here.

**Comment by 7iomka** [comment]
Same issue
@damirbogdanov your idea not working for me
I try to enable TAKE_LATEST behavior for all axios request
```ts

import type {
  AxiosInstance,
  AxiosResponse,
  CancelTokenSource,
  InternalAxiosRequestConfig,
} from 'axios';
import axios from 'axios';
import httpAdapter from 'axios/lib/adapters/http'; // << --the issue is here
import { appConfig } from '../config/app';

interface LatestRequest extends CancelTokenSource {
  cancel: () => void;
}

axios.defaults.adapter = httpAdapter;

let latestRequest: LatestRequest | null = null;

// default config
const instance: AxiosInstance = axios.create({
  baseURL: appConfig.APIBaseURL,
  withCredentials: true, // use it with proxy
});

instance.interceptors.request.use(
  (config: InternalAxiosRequestConfig) => {
    if (latestRequest) {
      latestRequest.cancel();
    }

    latestRequest = axios.CancelToken.source();

    config.cancelToken = latestRequest.token;

    return config;
  },
  (error) => {
    return Promise.reject(error);
  },
);

instance.interceptors.response.use(
  (response) => {
    return response;
  },
  (error) => {
    if (axios.isCancel(error)) {
      return Promise.resolve(null);
    }

    return Promise.reject(error);
  },
);
```

**Comment by arthurfiorette** [comment]
Hey @jasonsaayman. I know that importing code inside `lib` folder that are not part of the public API is strongly discouraged and it may break even in patch versions. 

The main reason we import things inside the `lib`/related folders is because we are trying to do something that axios does not support yet or reusing axios's logic in another part of our codebase.

The code is in OUR node_modules and just disallowing us from accessing it WILL not prevent us from using it. It is going to just create another bigger problem as, to solve our problem, we will copy/paste axios source code, which is much worse than our code breaking when upgrading axios, because it creates duplicity anyway and even future critical problems axios may fix won't reflect on our copy-pasted code.

And even if we achieve to import axios internal content without copy-pasting functions, I am sure that the code written to access it is more error-prone than just hypothetic incompatibilities axios may bring with a patch/minor version. 

I am strongly in favor of exporting everything axios has, even if it is going to be under a `internal`  property or similar names. 

I know that creating a issue to export `X` functionality will leave axios with much more control of what APIs you should care about or not, but in this issue/code/release meantime, the end-user will already have copied the function into his source code and probably will never look at it again, which leverages us to the same prior problem of axios fixing critical stuff and it not reflecting everywhere.

Happy to discuss more, also if you want, I can create another issue.

**Comment by jasonsaayman** [comment]
Ok fair enough, I have been gone for a bit of a hiatus, but seems as good as any time to start getting stuck in again. I will look into this

**Comment by trim21** [comment]
I have been using https://github.com/ds300/patch-package to patch package.json of axios for a while....

**Comment by DigitalBrainJS** [comment]
Perhaps we can export the lib folder with some prefix in the path to highlight this as an unsafe operation. I would like to avoid potential situations when we change the internal structure of Axios in the minor version, causing a bunch of issue reports. The next major version will have intensive changes to internal modules, as the core needs a deep rework in order to add middleware/hooks functionality and a plugin loader.

**Comment by marcin-wlodarczyk** [comment]
@DigitalBrainJS I'm trying to implement a custom Axios adapter.
Looks like the path `./unsafe/core/settle.js` is not recognized by TypeScript.

**Comment by eizodann** [comment]
> @DigitalBrainJS I'm trying to implement a custom Axios adapter. Looks like the path `./unsafe/core/settle.js` is not recognized by TypeScript.

Same here

**Comment by trim21** [comment]
> @DigitalBrainJS I'm trying to implement a custom Axios adapter. Looks like the path `./unsafe/core/settle.js` is not recognized by TypeScript.

axios internal files are plain js and not typed, you need to write your own type defs for this. 

There is nothing axios can do unless they rewrite whole project into TypeScript.
